File: Chris/2022-06-30-LongestCommonSubsequence.py
#Recursion => O((nm)^2) , space O(1)
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        if len(text1) == 1:
                return 1 if text1[0] in text2 else 0
        
        if len(text2) == 1:
                return 1 if text2[0] in text1 else 0
            
            
        
        if text1[0] == text2[0]:
            return 1 + self.longestCommonSubsequence(text1[1:], text2[1:])
        else:
            return max(self.longestCommonSubsequence(text1, text2[1:]), self.longestCommonSubsequence(text1[1:], text2))




# 2-D DP => TC : O(nm) SC: O(nm)

class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        
        lcs = [[0 for j in range(len(text2) + 1)] for i in range(len(text1) + 1)]
        
        for i in range(len(text1) -1, -1, -1):
            for j in range(len(text2) -1, -1, -1):
                if text1[i] == text2[j]:
                    lcs[i][j] = 1 + lcs[i+1][j+1]
                else:
                    lcs[i][j] = max(lcs[i][j+1], lcs[i+1][j])
                    
        
        return lcs[0][0]

File: Chloe Gwon/2022-06-30-Longest-Common-Subsequence.js
var longestCommonSubsequence = function(text1, text2) {
    let prevRow = Array(text2.length + 1).fill(0);
    for (var i=1; i<=text1.length; i++){
        const curRow = Array(text2.length+1).fill(0);
        for (var j=1; j<=text2.length; j++){
            if (text1[i-1] === text2[j-1]){
                curRow[j] = 1 + prevRow[j-1];
            }else{
                curRow[j] = Math.max(curRow[j-1], prevRow[j]);
            }
        }
        prevRow = curRow;
    }
    return prevRow[text2.length];
};
